src/library_tty.js

```diff
@@ -124,36 +123,36 @@ mergeInto(LibraryManager.library, {
         }
         return tty.input.shift();
       },
-      flush: function(tty) {
-        if (tty.output && tty.output.length > 0) {
-          Module['print'](tty.output.join(''));
-          tty.output = [];
-        }
-      },
       put_char: function(tty, val) {
         if (val === null || val === {{{ charCode('\n') }}}) {
-          Module['print'](tty.output.join(''));
+          Module['print'](UTF8ArrayToString(tty.output, 0));
           tty.output = [];
         } else {
-          tty.output.push(TTY.utf8.processCChar(val));
+          if (val != 0) tty.output.push(val); // val == 0 would cut text output off in the middle.
+        }
+      },
+      flush: function(tty) {
+        if (tty.output && tty.output.length > 0) {
+          Module['print'](UTF8ArrayToString(tty.output, 0));
```

Comment:
I am concerned about performance here. With this change, we move from a simple `join('')` to a loop that does `+= (char)`. On large strings it can create linear garbage, although JS engines might be able to use ropes to avoid that.